display_list/display_list_builder.h

```diff
@@ -376,65 +381,131 @@ class DisplayListBuilder final : public virtual Dispatcher,
     void add_compatible_op() {
-      if (!cannot_inherit_opacity) {
-        if (has_compatible_op) {
-          cannot_inherit_opacity = true;
+      if (!cannot_inherit_opacity_) {
+        if (has_compatible_op_) {
+          cannot_inherit_opacity_ = true;
         } else {
-          has_compatible_op = true;
+          has_compatible_op_ = true;
         }
       }
     }
+
+    // The filter to apply to the layer bounds when it is restored
+    std::shared_ptr<const DlImageFilter> filter() { return filter_; }
+
+    // is_unbounded should be set to true if we ever encounter an operation
+    // on a layer that either is unrestricted (|drawColor| or |drawPaint|)
+    // or cannot compute its bounds (some effects and filters) and there
+    // was no outstanding clip op at the time.
+    // When the layer is restored, the outer layer may then process this
+    // unbounded state by accumulating its own clip or transferring the
+    // unbounded state to its own outer layer.
+    // Typically the DisplayList will have been constructed with a cull
+    // rect which will act as a default clip for the outermost layer and
+    // the unbounded state of all sub layers will eventually be caught by
+    // that cull rect so that the overall unbounded state of the entire
+    // DisplayList will never be true.
+    //
+    // SkPicture treats these same conditions as a Nop (they accumulate
+    // the SkPicture cull rect, but if it was not specified then it is an
+    // empty Rect and so has no effect on the bounds).
+    // If the Calculator object accumulates this flag into the root layer,
+    // then at least we can make the caller aware of that exceptional
+    // condition via the |DisplayListBoundsCalculator::isUnbounded| call.
+    //
+    // Flutter is unlikely to ever run into this as the Dart mechanisms
+    // all supply a non-null cull rect for all Dart Picture objects,
+    // even if that cull rect is kGiantRect.
+    void set_unbounded() { is_unbounded_ = true; }
+
+    // |is_unbounded| should be called after |getLayerBounds| in case
+    // a problem was found during the computation of those bounds,
+    // the layer will have one last chance to flag an unbounded state.
+    bool is_unbounded() const { return is_unbounded_; }
+
+   private:
+    size_t save_layer_offset_;
+    bool has_layer_;
+    bool cannot_inherit_opacity_;
+    bool has_compatible_op_;
+    SkM44 matrix_;
+    SkMatrix matrix33_;
+    SkRect clip_bounds_;
+    std::shared_ptr<const DlImageFilter> filter_;
+    bool is_unbounded_;
+    bool has_deferred_save_op_ = false;
+
+    friend class DisplayListBuilder;
   };
 
   std::vector<LayerInfo> layer_stack_;
   LayerInfo* current_layer_;
+  RectBoundsAccumulator rect_bounds_accumulator_;
+  RTreeBoundsAccumulator rtree_bounds_accumulator_;
+  BoundsAccumulator* accumulator() {
+    if (need_produce_rtree_) {
+      return &rtree_bounds_accumulator_;
+    }
+    return &rect_bounds_accumulator_;
```

Comment:
Use ?: here? Or just have a pointer that points at the appropriate accumulator that is initialized in the constructor without having to figure it out each time?

Reply:
Also, below, if there was just a pointer for the accumulator in use, then the bounds and rtree methods could just return `accumulator->bounds()` or `accumulator->rtree()`

(RectAccumulator would implement the latter as `return nullptr` potentially with an FML_LOG(INFO))